### DISCUSSION TITLE
Camera parameter convert from Cesium

### DISCUSSION THREAD

**Discussion by gisjiashu** [iTwin/itwinjs-core#1181] [Q&A] [answered]
Hi ,

In Cesium app, the camera parameters contains Position(x, y, z), Heading, Pitch, and Roll.

I just found similiar function in iTwinjs which is LookAt funtion, but don't know how to convert Cesium parameters to iTwinjs.

Ps: I debug it using lookAtGlobalLocation(eyeHeight: number, pitchAngleRadians = 0, location?: GlobalLocation, eyePoint?: Point3d), but missing Heading and Roll parameters from Ceisum.

Thanks

**Comment by gisjiashu** [comment]
It seems that camrea need to rotate a angle. But failure while in setting ratotion in view.

**Comment by pmconne** [comment, marked as answer]
The iTwin.js equivalent of `Cesium.HeadingPitchRoll` is [YawPitchRollAngles](https://www.itwinjs.org/reference/geometry-core/cartesiangeometry/yawpitchrollangles/).
Here are [diagrams](https://www.itwinjs.org/learning/frontend/views/#viewstate-parameters) illustrating iTwin.js camera parameters.

**Reply by gisjiashu** [reply]
Is that means excute before viewState.lookAtUsingLensAngle function, should set Rotation first?
The rotation parameter comes from YawPitchRollAngles propoerty.

**Reply by bbastings** [reply]
lookAtUsingLensAngle will set the view rotation. The vector from targetPoint to eyePoint becomes view Z, and the upVector determines view Y. Then Y cross Z becomes View X.

YawPitchRollAngles.toMatrix3d could be used if you were going to call Viewport.setRotation and set up all the view parameters yourself according to [diagram](https://www.itwinjs.org/learning/frontend/views/#viewstate-parameters), not if you're going to call lookAtUsingLensAngle.

**Reply by gisjiashu** [reply]
In the source code, we just used the pitch angle parameter from Cesium. Cann't set heading angle in the lookAtUsingLensAngle. 
Is the vector from targetPoint to eyePoint(iTwin.js) means Heading(Cesium) angle?

**Reply by bbastings** [reply]
> Is the vector from targetPoint to eyePoint(iTwin.js) means Heading(Cesium) angle?

They aren't directly comparable, one is a Vector3d and the other is an angle in radians...

You could do something like HeadingPitchRoll -> YawPitchRoll -> toMatrix3d, and use Matrix3d.getColumn(2) to compute a targetPoint, etc.
